docs/design/validation.md

```diff
@@ -0,0 +1,2048 @@
+
+Attribute value validations should be implemented as a slice of the interface type on `schema.Attribute`. Multiple attribute validation, such as declaring conflicting attributes on attributes themselves, should be implemented as a separate slice of that differing interface type on `schema.Attribute`. This would be in addition to supporting that functionality with resource level multiple attribute validation.
+
+Attribute value validations should be required to accept the attribute path in its native type as a parameter. This will allow a flexible implementation for provider developers that may desire advanced logic based on the path.
+
+Validation functions should be required to return framework defined error types. These errors will either result in the framework returning consistent error diagnostics or callers (such as wrapper validation functions) otherwise handling these results in a predictable manner. Error types that equate to consistent warning diagnostics can also be provided, if desired.
+
+Resource level multiple attribute validation functions should be implemented separately from plan modifications to separate concerns. For example:
+
+### Attribute Level Example Implementation
+
+Example framework code:
+
+```go
+// Well defined error types
+type ValueValidatorError interface {}
+
+type ValueValidatorInvalidTypeError interface {
+    ValueValidatorError
+}
+
+type ValueValidatorUnknownValueError interface {
+    ValueValidatorError
+}
+
+type ValueValidatorUnsuccessfulValidationError interface {
+    ValueValidatorError
+}
+
+// ValueValidator is an interface type for implementing common validation functionality.
+type ValueValidator interface {
+    Description(context.Context) string
+    MarkdownDescription(context.Context) string
+}
+
+// ValueValidators is a type alias for a slice of ValueValidator.
+type ValueValidators []ValueValidator
+
+// Descriptions returns all ValueValidator Description
+func (vs ValueValidators) Descriptions(ctx context.Context) []string {
+    // ...
+}
+
+// MarkdownDescriptions returns all ValueValidator MarkdownDescription
+func (vs ValueValidators) MarkdownDescriptions(ctx context.Context) []string {
+    // ...
+}
+
+// Validates performs all ValueValidator Validate or ValidateWithProvider
+func (vs ValueValidators) Validates(ctx context.Context) diag.Diagnostics {
+    // ...
+}
+
+// GenericValueValidator describes value validation without a strong type.
+//
+// While it is generally preferred to use the typed validation interfaces,
+// such as StringValueValidator, this interface allows custom implementations
+// where the others may not be suitable. The Validate function is responsible
+// for protecting against attr.Value type assertion panics.
+type GenericValueValidator interface {
+    ValueValidator
+    Validate(context.Context, *tftypes.AttributePath, attr.Value) error
+}
+
+// StringValueValidator is an interface type for implementing String value validation.
```

Comment:
`StringValueValidator` has great UX - no type assertions needed in the provider code!

How would this work with complex types, e.g. `types.List`? Since there are as many list types as there are element types, it seems that the options are either:
1. Create a single `ListValueValidator` interface, and require type assertions on the list elements inside the `Validate` function,
2. Create multiple `ListOfStringsValueValidator`, etc, interfaces, accepting that not all cases can be covered but trying to provide helpers for the most common ones, or
3. Do not create strongly typed validator interfaces for complex types.

Reply:
In my head I was thinking in the area of 2 and leaving an escape hatch interface (3). 1 could make 2 slightly easier to do, so might be beneficial.

I'm also curious though of what validation space exists here -- in the previous framework there is not much available validation for complex types (e.g. you can only validate the number of elements in a list/set/map). There have been very odd cases of needing a list type for ordering, but needing element uniqueness (like a set).

I think my focus would be covering the very common string/number cases then allowing those more odd cases to bubble up.